relaymgr/relaymgr.go

```diff
@@ -0,0 +1,138 @@
+package relaymgr
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/multiformats/go-multiaddr"
+	logger "github.com/textileio/go-log/v2"
+)
+
+const (
+	connProtectTag = "auction-relay"
+)
+
+var (
+	log = logger.Logger("relaymgr")
+
+	pollFrequency = time.Second * 10
+)
+
+// RelayManager connects a libp2p host to an external relay and do a best-effort
+// in keeping the connection healthy.
+type RelayManager struct {
+	host        host.Host
+	relayAddr   peer.AddrInfo
+	connNotifee *connNotifee
+
+	closeOnce   sync.Once
+	closeCtx    context.Context
+	closeSignal context.CancelFunc
+	closed      chan struct{}
+}
+
+// New connects the provided host to the remote relay, doing a best-effort
+// in keeping the connection healthy. The provided context is only used for
+// the initial connection to the relay. To shutdown call Close().
+func New(ctx context.Context, h host.Host, relayMultiaddress string) (*RelayManager, error) {
+	relayAddr, err := multiaddr.NewMultiaddr(relayMultiaddress)
+	if err != nil {
+		return nil, fmt.Errorf("parsing relay multiaddr: %s", err)
+	}
+	addrInfo, err := peer.AddrInfoFromP2pAddr(relayAddr)
+	if err != nil {
+		return nil, fmt.Errorf("get addr-info from relay multiaddr: %s", err)
+	}
+
+	closeCtx, closeSignal := context.WithCancel(context.Background())
+	rm := &RelayManager{
+		host:      h,
+		relayAddr: *addrInfo,
+
+		closeCtx:    closeCtx,
+		closeSignal: closeSignal,
+		closed:      make(chan struct{}),
+	}
+	rm.connNotifee = &connNotifee{rm: rm}
+	h.Network().Notify(rm.connNotifee)
+
+	if err := rm.connect(); err != nil {
+		return nil, fmt.Errorf("connecting to relay: %s", err)
+	}
+
+	go rm.keepHealthy()
+
+	return rm, nil
+}
+
+// Close stops relay manager work to keep a healthy connection with the relay.
+func (rm *RelayManager) Close() error {
+	rm.closeOnce.Do(func() {
+		log.Infof("closing relay manager")
+
+		rm.host.ConnManager().Unprotect(rm.relayAddr.ID, connProtectTag)
+		rm.host.Network().StopNotify(rm.connNotifee)
+		rm.closeSignal()
+		<-rm.closed
+		log.Infof("relay manager closed")
+	})
+	return nil
+}
+
+func (rm *RelayManager) keepHealthy() {
```

Comment:
We have two mechanisms for checking healthiness.
This one is simply polling every X seconds to see if we still have an active connection with the relay. 
If that isn't the case, it will reconnect automatically.